## Test Run 1: Load AbstractNet Dataset: 70K unlabeled (and 2K labeled) Abstracts into DB

This notebook loads the dataset and create labeled *candidates* through labeling function. Feel extra free to document/bring up any upcoming confusion throughout the test run, e.g. are the following two consistent, the **(segment, label) pair** that we want to have, and the **candidates** that we instruct snorkel to extract? 

Before everything, please ensure that you have followed project-level ``README.md`` and installed all python dependencies, e.g. ``tika``.  

We filtered out null abstracts from `ClydeDB.csv` ([AbstractSegmentationCrowdNLP Git repo](https://github.com/zhoujieli/AbstractSegmentationCrowdNLP.git)), resulting in 48,914 valid ones out of 56,851 total abstracts. The 48,914 abstracts are saved to `data/70kpaper.tsv`.



In this section, we preprocess documents by parsing them into *contexts*. *Candidates* are extracted out of *contexts*, which are *instances* (one of the *background*, *purpose*, *mechanism*, *method*, and *findings*).

In [1]:
%load_ext autoreload
%autoreload 2
%matplotlib inline
import os

from snorkel import SnorkelSession
session = SnorkelSession()

# # Here, we just set how many documents we'll process for automatic testing- you can safely ignore this!
n_docs = 500 if 'CI' in os.environ else 7000 #  change the number 1000 to 60,000 for real dataset 

from snorkel.parser import TSVDocPreprocessor

doc_preprocessor = TSVDocPreprocessor('data/70kpaper_061418_cleaned_noBookLecture.tsv', encoding="utf-8",max_docs=n_docs)

Get statistics on the number of documents and sentences, as below. This could take 5-8 minutes to load ~60K papers (see progress bar, also might have exception). The following code parses docs into sentences by period, averaging 4.49 sentences per documents. Earlier I spent a few hours debugging some hidden formatting error that confuses Spacey. Need to ensure that we format raw data from .csv into .tsv *without* preceeding and appending quotes.  


In [2]:
from snorkel.parser.spacy_parser import Spacy
from snorkel.parser import CorpusParser


corpus_parser = CorpusParser(parser=Spacy())
%time corpus_parser.apply(doc_preprocessor, count=n_docs)


from snorkel.models import Document, Sentence  # defined in context.py file

print("Documents:", session.query(Document).count())
print("Sentences:", session.query(Sentence).count())

Clearing existing...
Running UDF...
[===                                     ] 6%Documents: 0
Sentences: 0



KeyboardInterrupt



Next we extract `candidates` by defining the specific `CandidateExtractor` for abstract segmentation. We take *Background* as an example and come back with other segmentation parts, i.e., *mechanism*, *method*, *findings*, later. 

Some more explanation based on my understanding: 
    
1. `Candidates` are defined as a class that contains 1+ `Span` objects within one `Sentence` context.  
    
2. `Span(s)` correspond to conceptual categories in text like people or disease names. 

To visualize, here is the Context Hierarchy 
<img src="../../tutorials/workshop/imgs/context-hierarchy.jpg" width="300px;">

All `Context(s)` are hierarchical in Snorkel. The default objects provided by Snorkel are show above. 
    
In the intro tutorial example, their `candidate` represents the possible `Spouse` mention `(Barrack Obama, Michelle Obama)`. As readers, we know this mention is true due to external knowledge and the keyword of `wedding` occuring later in the sentence. (Reference: (1) section `Writing a basic CandidateExtractor` in [Intro_tutorial_1](../intro/Intro_Tutorial_1.ipynb); (2) section `Candidate Member Functions and Variables` in [Workshop_1_Snorkel_API](../workshop/Workshop_1_Snorkel_API.ipynb)) 


+ Background: 
  - "Recent research ... ", 
  - "... have/has been widely ...", 
  - "How ... ?" (and as the first sentence), 
  - "Previous work...", 
  - "Motivated by...", 
  - "The success of ...", etc.
+ Mechanism:
  - something
  - some other pattern

We define `CandidateExtractor` as a wrapper of `CandidateSpace` (e.g. `Ngrams` is one type of `CandidateSpace`) and `Matcher` (e.g. `DictionaryMatcher`, `PersonMatcher`). Please make sure that `longest_match_only=True`, since this gets us longest span that contains dictionary words. (Reference: source code [candidates.py](../../snorkel/candidates.py) and [matchers.py](../../snorkel/matchers.py)]

Get the longest sentence length. This value is important, to be used soon.

In [4]:
sents = session.query(Sentence).all()
n_max_corpus=0
for sent in sents:
    n_max_corpus=max(n_max_corpus,len(sent.words))

print("The longest sentence has "+str(n_max_corpus)+" tokens.")

The longest sentence has 101 tokens.


In [7]:
from snorkel.models import candidate_subclass
from snorkel.candidates import Ngrams, CandidateExtractor
from snorkel.matchers import PersonMatcher,DictionaryMatch

Background = candidate_subclass('Background', ['background_cue'])

ngrams = Ngrams(n_max=n_max_corpus) # we define the maximum n value as n_max_corpus
# Start simple: any ngram that matches the dictionary are *background* candidates! 
dict_matcher=DictionaryMatch(d=['previous','motivated','recent','widely'],longest_match_only=True) 
cand_extractor=CandidateExtractor(Background, [ngrams], [dict_matcher])

Now we apply the defined `CandidateExtractor` to the all `Sentences` in the collection (splitted 90/10/10 for train/dev/test). 

In [8]:
from snorkel.models import Document
from util import number_of_people

docs = session.query(Document).order_by(Document.name).all()

train_sents = set()
dev_sents   = set()
test_sents  = set()

for i, doc in enumerate(docs):
    for s in doc.sentences:
#         if number_of_people(s) <= 5:
            if i % 10 == 8:
                
                dev_sents.add(s)
            elif i % 10 == 9:
                test_sents.add(s)
            else:
                train_sents.add(s)
                
for i, sents in enumerate([train_sents, dev_sents, test_sents]):
    %time cand_extractor.apply(sents, split=i)
    print("Split "+str(i)+" - number of candidates extracted:", session.query(Background).filter(Background.split == i).count(),"\n\n")
            

Clearing existing...
Running UDF...
[========================================] 100%

CPU times: user 7.51 s, sys: 193 ms, total: 7.71 s
Wall time: 7.6 s
Split 0 - number of candidates extracted: 104 


Clearing existing...
Running UDF...
[========================================] 100%

CPU times: user 1.23 s, sys: 87.4 ms, total: 1.32 s
Wall time: 1.27 s
Split 1 - number of candidates extracted: 11 


Clearing existing...
Running UDF...
[========================================] 100%

CPU times: user 1.22 s, sys: 82.5 ms, total: 1.3 s
Wall time: 1.26 s
Split 2 - number of candidates extracted: 9 




Let's take a look at a few of those extracted `Candidates`! Obviously, since we used `DictionaryMatcher`, all `Candidates` will contain at least one word from the set `['previous','motivated','recent','widely']`.

In [18]:
from IPython.display import Markdown, display
def printmd(string):
    display(Markdown(string))
    
cands = session.query(Background).filter(Background.split == 0).all()

document_list=list()
for i in range(len(cands)):
    printmd("**"+str(i)+"/"+str(len(cands))+" Candidate/Span:**\t`"+str(cands[i].background_cue)+"`")
    printmd("**Its parent Sentence's text:**\t"+str(cands[i].get_parent().text))
    document_list+=[cands[i].background_cue.sentence]
    print()
    
print("Check if any of the extracted candidates come from the same sentence? "+str(len(set(document_list))!=len(document_list)))

**0/104 Candidate/Span:**	`Span("b'Recently, type systems based on constructive modal logic have been proposed as an expressive \\nbasis for run-time code generation [DP96,WLPD98], partial evaluation [Dav96], and general \\nmeta-programming [MTBS99,DPS97].'", sentence=3237, chars=[0,218], words=[0,40])`

**Its parent Sentence's text:**	Recently, type systems based on constructive modal logic have been proposed as an expressive \nbasis for run-time code generation [DP96,WLPD98], partial evaluation [Dav96], and general \nmeta-programming [MTBS99,DPS97].

**1/104 Candidate/Span:**	`Span("b'Abstract: Active physiological tremor compensation instruments have been under research and development recently.'", sentence=4785, chars=[0,112], words=[0,14])`

**Its parent Sentence's text:**	Abstract: Active physiological tremor compensation instruments have been under research and development recently.

**2/104 Candidate/Span:**	`Span("b'Recently, there has been considerable interest in the use of Model Checking for Systems Biology.'", sentence=1771, chars=[0,95], words=[0,16])`

**Its parent Sentence's text:**	Recently, there has been considerable interest in the use of Model Checking for Systems Biology.

**3/104 Candidate/Span:**	`Span("b'Mechanism design is the art of designing the rules of the game so that the agents are motivated to report their preferences truthfully, and a desirable outcome is chosen.'", sentence=3177, chars=[0,169], words=[0,30])`

**Its parent Sentence's text:**	Mechanism design is the art of designing the rules of the game so that the agents are motivated to report their preferences truthfully, and a desirable outcome is chosen.

**4/104 Candidate/Span:**	`Span("b'This claim is supported by recent results in music vs. speech classification, structure from sound, robust music identification and sound object recognition.'", sentence=5220, chars=[0,156], words=[0,24])`

**Its parent Sentence's text:**	This claim is supported by recent results in music vs. speech classification, structure from sound, robust music identification and sound object recognition.

**5/104 Candidate/Span:**	`Span("b'Abstract A recent trend (especially in electronic commerce) is higher levels of expressiveness in the mechanisms that mediate interactions such as auctions, exchanges, catalog offers, voting systems, matching of peers, and so on.'", sentence=1776, chars=[0,228], words=[0,40])`

**Its parent Sentence's text:**	Abstract A recent trend (especially in electronic commerce) is higher levels of expressiveness in the mechanisms that mediate interactions such as auctions, exchanges, catalog offers, voting systems, matching of peers, and so on.

**6/104 Candidate/Span:**	`Span("b'Coordinated Sampling (cSamp) is a recent proposal for improving the flow monitoring capabilities of ISPs to address these demands.'", sentence=4142, chars=[0,129], words=[0,21])`

**Its parent Sentence's text:**	Coordinated Sampling (cSamp) is a recent proposal for improving the flow monitoring capabilities of ISPs to address these demands.

**7/104 Candidate/Span:**	`Span("b'Abstract Information-Centric Networking (ICN) has seen a significant resurgence in recent years.'", sentence=4921, chars=[0,95], words=[0,16])`

**Its parent Sentence's text:**	Abstract Information-Centric Networking (ICN) has seen a significant resurgence in recent years.

**8/104 Candidate/Span:**	`Span("b'Abstract Much recent attention, both experimental and theoretical, has been focussed on classification algorithms which produce voted combinations of classifiers.'", sentence=3778, chars=[0,161], words=[0,22])`

**Its parent Sentence's text:**	Abstract Much recent attention, both experimental and theoretical, has been focussed on classification algorithms which produce voted combinations of classifiers.

**9/104 Candidate/Span:**	`Span("b'Recent theoretical work has shown that the impressive generalization performance of algorithms like AdaBoost can be attributed to the classifier having large margins on the training data.'", sentence=3779, chars=[0,186], words=[0,27])`

**Its parent Sentence's text:**	Recent theoretical work has shown that the impressive generalization performance of algorithms like AdaBoost can be attributed to the classifier having large margins on the training data.

**10/104 Candidate/Span:**	`Span("b'In previous work we have developed an optimal method D* to plan paths when the environment is not known ahead of time, but, rather is discovered as the robot moves around.'", sentence=4492, chars=[0,170], words=[0,34])`

**Its parent Sentence's text:**	In previous work we have developed an optimal method D* to plan paths when the environment is not known ahead of time, but, rather is discovered as the robot moves around.

**11/104 Candidate/Span:**	`Span("b'Recent proposals have advocated the use of consolidation of \n'", sentence=2205, chars=[0,60], words=[0,9])`

**Its parent Sentence's text:**	Recent proposals have advocated the use of consolidation of 


**12/104 Candidate/Span:**	`Span("b'We investigate this question for a space of matrix decomposition models which can express a variety of widely used models from unsupervised learning.'", sentence=2714, chars=[0,148], words=[0,23])`

**Its parent Sentence's text:**	We investigate this question for a space of matrix decomposition models which can express a variety of widely used models from unsupervised learning.

**13/104 Candidate/Span:**	`Span("b'We discuss, analyze, and experiment with a setup motivated by the behavior of real-world distributed computation networks, where the machines are differently slow at different time.'", sentence=1266, chars=[0,180], words=[0,31])`

**Its parent Sentence's text:**	We discuss, analyze, and experiment with a setup motivated by the behavior of real-world distributed computation networks, where the machines are differently slow at different time.

**14/104 Candidate/Span:**	`Span("b'The recent spate of security issues and allegations of \xe2\x80\x9clost votes\xe2\x80\x9d in the US demonstrates the inadequacy of the standards used to evaluate our election systems.'", sentence=2481, chars=[0,160], words=[0,28])`

**Its parent Sentence's text:**	The recent spate of security issues and allegations of “lost votes” in the US demonstrates the inadequacy of the standards used to evaluate our election systems.

**15/104 Candidate/Span:**	`Span("b'To do so, we make use of a recently developed fine-grained taxonomy of human-object grasps.'", sentence=1887, chars=[0,90], words=[0,20])`

**Its parent Sentence's text:**	To do so, we make use of a recently developed fine-grained taxonomy of human-object grasps.

**16/104 Candidate/Span:**	`Span("b'While recent work in network verification has made giant strides to reduce this effort, they focus on simple reachability properties and cannot handle context-dependent policies (eg, how many connections has a host spawned) that operators realize using stateful network functions (NFs).'", sentence=4155, chars=[0,285], words=[0,50])`

**Its parent Sentence's text:**	While recent work in network verification has made giant strides to reduce this effort, they focus on simple reachability properties and cannot handle context-dependent policies (eg, how many connections has a host spawned) that operators realize using stateful network functions (NFs).

**17/104 Candidate/Span:**	`Span("b'Abstract Previous research on automated mechanism design (proposed in UAI-02) assumed that the outcome space was flatly represented, which makes that work inapplicable if the outcome space is exponential, as it is, for example, in multi-item auctions.'", sentence=1767, chars=[0,250], words=[0,45])`

**Its parent Sentence's text:**	Abstract Previous research on automated mechanism design (proposed in UAI-02) assumed that the outcome space was flatly represented, which makes that work inapplicable if the outcome space is exponential, as it is, for example, in multi-item auctions.

**18/104 Candidate/Span:**	`Span("b'During the seminar, several participants presented their recent research, and ongoing work and open problems were discussed.'", sentence=3324, chars=[0,123], words=[0,19])`

**Its parent Sentence's text:**	During the seminar, several participants presented their recent research, and ongoing work and open problems were discussed.

**19/104 Candidate/Span:**	`Span("b'There have been important recent game-theoretic analyses of spiteful bidding assuming all agents are equally spiteful.'", sentence=3993, chars=[0,117], words=[0,18])`

**Its parent Sentence's text:**	There have been important recent game-theoretic analyses of spiteful bidding assuming all agents are equally spiteful.

**20/104 Candidate/Span:**	`Span("b'From the graph we compute strategies using previous work \n'", sentence=2366, chars=[0,57], words=[0,9])`

**Its parent Sentence's text:**	From the graph we compute strategies using previous work 


**21/104 Candidate/Span:**	`Span("b"Abstract Recent trends have exposed three key problems in today's operating systems."", sentence=1722, chars=[0,83], words=[0,13])`

**Its parent Sentence's text:**	Abstract Recent trends have exposed three key problems in today's operating systems.

**22/104 Candidate/Span:**	`Span("b'Within the sciences of the mind, issues of material embodiment and environmental embedding have emerged as important areas of research attention over recent decades.'", sentence=4295, chars=[0,164], words=[0,25])`

**Its parent Sentence's text:**	Within the sciences of the mind, issues of material embodiment and environmental embedding have emerged as important areas of research attention over recent decades.

**23/104 Candidate/Span:**	`Span("b'Previous Work: Highway Control'", sentence=1229, chars=[0,29], words=[0,4])`

**Its parent Sentence's text:**	Previous Work: Highway Control

**24/104 Candidate/Span:**	`Span("b'There has been recently a reawaking of interest in many aspects of realizability interpretations---especially as regards semantics of type theories for constructive reasoning and semantics of programming languages.'", sentence=4300, chars=[0,213], words=[0,30])`

**Its parent Sentence's text:**	There has been recently a reawaking of interest in many aspects of realizability interpretations---especially as regards semantics of type theories for constructive reasoning and semantics of programming languages.

**25/104 Candidate/Span:**	`Span("b'Previous approaches such as ridge regression, support vector methods, and regularization networks are included as special cases.'", sentence=3807, chars=[0,127], words=[0,19])`

**Its parent Sentence's text:**	Previous approaches such as ridge regression, support vector methods, and regularization networks are included as special cases.

**26/104 Candidate/Span:**	`Span("b'Abstract: The recent proliferation of richly structured probabilistic models raises the question of how to automatically determine an appropriate model for a dataset.'", sentence=2713, chars=[0,165], words=[0,24])`

**Its parent Sentence's text:**	Abstract: The recent proliferation of richly structured probabilistic models raises the question of how to automatically determine an appropriate model for a dataset.

**27/104 Candidate/Span:**	`Span("b'In our opinion, the literature has, until quite recently, placed too much emphasis on probabilistic inference machinery, while paying \n'", sentence=3419, chars=[0,134], words=[0,23])`

**Its parent Sentence's text:**	In our opinion, the literature has, until quite recently, placed too much emphasis on probabilistic inference machinery, while paying 


**28/104 Candidate/Span:**	`Span("b'Previous work with the previous generation of mobile phones has shown that such an assumption is false.'", sentence=1814, chars=[0,102], words=[0,17])`

**Its parent Sentence's text:**	Previous work with the previous generation of mobile phones has shown that such an assumption is false.

**29/104 Candidate/Span:**	`Span("b'ABSTRACT Motivated by many practical applications that have to compute in the presence of uncertainty, we propose a monadic probabilistic language based upon the mathematical notion of sampling function.'", sentence=3160, chars=[0,202], words=[0,30])`

**Its parent Sentence's text:**	ABSTRACT Motivated by many practical applications that have to compute in the presence of uncertainty, we propose a monadic probabilistic language based upon the mathematical notion of sampling function.

**30/104 Candidate/Span:**	`Span("b'Recent research efforts have created tools that automatically localize the problem to a small number of potential culprits, but research is needed to understand what visualization techniques work best for helping distributed systems developers understand \n'", sentence=4670, chars=[0,255], words=[0,36])`

**Its parent Sentence's text:**	Recent research efforts have created tools that automatically localize the problem to a small number of potential culprits, but research is needed to understand what visualization techniques work best for helping distributed systems developers understand 


**31/104 Candidate/Span:**	`Span("b'To this end, we develop a logically motivated theory of parametric polymorphism, reminiscent of the Girard-Reynolds polymorphic \xce\xbb-calculus, but casted in the setting of concurrent processes.'", sentence=3939, chars=[0,189], words=[0,31])`

**Its parent Sentence's text:**	To this end, we develop a logically motivated theory of parametric polymorphism, reminiscent of the Girard-Reynolds polymorphic λ-calculus, but casted in the setting of concurrent processes.

**32/104 Candidate/Span:**	`Span("b"Abstract We present recent advances from our efforts in increasing coverage, robustness, generality and speed of JANUS, CMU's speech-to-speech translation system."", sentence=2885, chars=[0,161], words=[0,29])`

**Its parent Sentence's text:**	Abstract We present recent advances from our efforts in increasing coverage, robustness, generality and speed of JANUS, CMU's speech-to-speech translation system.

**33/104 Candidate/Span:**	`Span("b'To circumvent this impossibility result, recent research has investigated \n'", sentence=2575, chars=[0,74], words=[0,10])`

**Its parent Sentence's text:**	To circumvent this impossibility result, recent research has investigated 


**34/104 Candidate/Span:**	`Span("b'Abstract: Recent work in machine learning has significantly benefited semantic extraction tasks in computer vision, particularly for object recognition and image retrieval.'", sentence=5218, chars=[0,171], words=[0,24])`

**Its parent Sentence's text:**	Abstract: Recent work in machine learning has significantly benefited semantic extraction tasks in computer vision, particularly for object recognition and image retrieval.

**35/104 Candidate/Span:**	`Span("b'In previous work, Polakow presented ordered linear logic and gave a backward-chaining operational semantics semantics for the uniform fragment of the logic, and the authors have previously presented a fragment of ordered linear logic suitable to a forward-chaining operational semantics.'", sentence=2969, chars=[0,286], words=[0,46])`

**Its parent Sentence's text:**	In previous work, Polakow presented ordered linear logic and gave a backward-chaining operational semantics semantics for the uniform fragment of the logic, and the authors have previously presented a fragment of ordered linear logic suitable to a forward-chaining operational semantics.

**36/104 Candidate/Span:**	`Span("b'Due to the needs of many real-world applications, in recent years an increasing amount of reserch has been devoted to machine learning on relational data with more complex structure.'", sentence=2291, chars=[0,181], words=[0,32])`

**Its parent Sentence's text:**	Due to the needs of many real-world applications, in recent years an increasing amount of reserch has been devoted to machine learning on relational data with more complex structure.

**37/104 Candidate/Span:**	`Span("b'Machine learning develops intelligent computer systems that are able to generalizefrom previously seen examples.'", sentence=3747, chars=[0,111], words=[0,14])`

**Its parent Sentence's text:**	Machine learning develops intelligent computer systems that are able to generalizefrom previously seen examples.

**38/104 Candidate/Span:**	`Span("b'As a result of these changes, the previous components must become interoperable with the \n'", sentence=1969, chars=[0,89], words=[0,15])`

**Its parent Sentence's text:**	As a result of these changes, the previous components must become interoperable with the 


**39/104 Candidate/Span:**	`Span("b'It is based on a recently discovered connection between homotopy the-ory and type theory.'", sentence=4352, chars=[0,88], words=[0,16])`

**Its parent Sentence's text:**	It is based on a recently discovered connection between homotopy the-ory and type theory.

**40/104 Candidate/Span:**	`Span("b'Previous versions of the robot have been remotely actuated through push-pull wires, while visual feedback was provided by fiber optic transmission.'", sentence=3012, chars=[0,146], words=[0,24])`

**Its parent Sentence's text:**	Previous versions of the robot have been remotely actuated through push-pull wires, while visual feedback was provided by fiber optic transmission.

**41/104 Candidate/Span:**	`Span("b'In this lecture we combine ideas from the previous two lectures, linear monadic logic programming and higher-order abstract syntax, to present a specification technique for programming languages we call substructural operational semantics.'", sentence=2622, chars=[0,238], words=[0,36])`

**Its parent Sentence's text:**	In this lecture we combine ideas from the previous two lectures, linear monadic logic programming and higher-order abstract syntax, to present a specification technique for programming languages we call substructural operational semantics.

**42/104 Candidate/Span:**	`Span("b'Marian ran a straw poll on that question at ICSE, at a recent social gathering, \n'", sentence=4350, chars=[0,80], words=[0,17])`

**Its parent Sentence's text:**	Marian ran a straw poll on that question at ICSE, at a recent social gathering, 


**43/104 Candidate/Span:**	`Span("b'Abstract Convolutional neural nets (CNNs) have demonstrated remarkable performance in recent history.'", sentence=4022, chars=[0,100], words=[0,14])`

**Its parent Sentence's text:**	Abstract Convolutional neural nets (CNNs) have demonstrated remarkable performance in recent history.

**44/104 Candidate/Span:**	`Span("b'This paper builds off of recent work on rapidly exponentially stabilizing control Lyapunov functions (RES-CLF) and control Lyapunov function based quadratic programs (CLFQP) for underactuated hybrid systems.'", sentence=3535, chars=[0,206], words=[0,33])`

**Its parent Sentence's text:**	This paper builds off of recent work on rapidly exponentially stabilizing control Lyapunov functions (RES-CLF) and control Lyapunov function based quadratic programs (CLFQP) for underactuated hybrid systems.

**45/104 Candidate/Span:**	`Span("b'each Mars Exploration Rover (MER) affords better obstacle avoidance than does the previous version.'", sentence=4895, chars=[0,98], words=[0,16])`

**Its parent Sentence's text:**	each Mars Exploration Rover (MER) affords better obstacle avoidance than does the previous version.

**46/104 Candidate/Span:**	`Span("b'Abstract Motivated by a radically new peer review system that the National Science Foundation recently experimented with, we study peer review systems in which proposals are reviewed by PIs who have submitted proposals themselves.'", sentence=3658, chars=[0,229], words=[0,35])`

**Its parent Sentence's text:**	Abstract Motivated by a radically new peer review system that the National Science Foundation recently experimented with, we study peer review systems in which proposals are reviewed by PIs who have submitted proposals themselves.

**47/104 Candidate/Span:**	`Span("b'In previous work we have shown how range readings from radio tags placed in the environment can be used to localize a robot.'", sentence=4459, chars=[0,123], words=[0,23])`

**Its parent Sentence's text:**	In previous work we have shown how range readings from radio tags placed in the environment can be used to localize a robot.

**48/104 Candidate/Span:**	`Span("b'Kidney exchanges represent a truly fielded example of technology at the intersection of artificial intelligence and economics\xe2\x80\x94an example that is both recent and an active research area.'", sentence=3478, chars=[0,184], words=[0,29])`

**Its parent Sentence's text:**	Kidney exchanges represent a truly fielded example of technology at the intersection of artificial intelligence and economics—an example that is both recent and an active research area.

**49/104 Candidate/Span:**	`Span("b'The Urban Challenge represents a technological leap beyond the previous Grand Challenges.'", sentence=3143, chars=[0,88], words=[0,12])`

**Its parent Sentence's text:**	The Urban Challenge represents a technological leap beyond the previous Grand Challenges.

**50/104 Candidate/Span:**	`Span("b'In this paper, we present recent results with using range from radio for mobile robot localization.'", sentence=4458, chars=[0,98], words=[0,17])`

**Its parent Sentence's text:**	In this paper, we present recent results with using range from radio for mobile robot localization.

**51/104 Candidate/Span:**	`Span("b'We recently proposed an approach---called automated mechanism design---where a mechanism is computed for the preference \n'", sentence=3178, chars=[0,120], words=[0,19])`

**Its parent Sentence's text:**	We recently proposed an approach---called automated mechanism design---where a mechanism is computed for the preference 


**52/104 Candidate/Span:**	`Span("b'Abstract Automated abstraction algorithms for sequential imperfect information games have recently emerged as a key component in developing competitive game theory-based agents.'", sentence=4594, chars=[0,176], words=[0,24])`

**Its parent Sentence's text:**	Abstract Automated abstraction algorithms for sequential imperfect information games have recently emerged as a key component in developing competitive game theory-based agents.

**53/104 Candidate/Span:**	`Span("b'A recent trend towards greater automation of earthmoving machines, such as backhoes, loaders, and dozers, reflects a larger movement in the construction industry to improve productivity, efficiency, and safety.'", sentence=2180, chars=[0,209], words=[0,35])`

**Its parent Sentence's text:**	A recent trend towards greater automation of earthmoving machines, such as backhoes, loaders, and dozers, reflects a larger movement in the construction industry to improve productivity, efficiency, and safety.

**54/104 Candidate/Span:**	`Span("b'Despite the proliferation of detection and containment techniques in the worm defense literature, simple threshold-based methods remain the most widely deployed and most popular approach among practitioners.'", sentence=2300, chars=[0,206], words=[0,30])`

**Its parent Sentence's text:**	Despite the proliferation of detection and containment techniques in the worm defense literature, simple threshold-based methods remain the most widely deployed and most popular approach among practitioners.

**55/104 Candidate/Span:**	`Span("b'While previous studies demonstrate the feasibility of using ACT-R to model collective cognition, as well as sensemaking processes at the individual level, the \n'", sentence=4273, chars=[0,159], words=[0,27])`

**Its parent Sentence's text:**	While previous studies demonstrate the feasibility of using ACT-R to model collective cognition, as well as sensemaking processes at the individual level, the 


**56/104 Candidate/Span:**	`Span("b'The presentation from the previous lecture would have been slightly simpler if we had presupposed an explicit conjunction form for programs where each inference rule has exactly one premiss.'", sentence=4850, chars=[0,189], words=[0,29])`

**Its parent Sentence's text:**	The presentation from the previous lecture would have been slightly simpler if we had presupposed an explicit conjunction form for programs where each inference rule has exactly one premiss.

**57/104 Candidate/Span:**	`Span("b'In this paper, motivated by modeling of conflict scenarios in societies with multiple ethno-\n'", sentence=4103, chars=[0,92], words=[0,15])`

**Its parent Sentence's text:**	In this paper, motivated by modeling of conflict scenarios in societies with multiple ethno-


**58/104 Candidate/Span:**	`Span("b'Session types are widely accepted as an expressive discipline for structuring communications in concurrent and distributed systems.'", sentence=1404, chars=[0,130], words=[0,17])`

**Its parent Sentence's text:**	Session types are widely accepted as an expressive discipline for structuring communications in concurrent and distributed systems.

**59/104 Candidate/Span:**	`Span("b'Previously, some basic machine learning classification algorithms have been used to evaluate the diagnostic power of noted antigens, but the issue of which algorithms perform best on this type of data is \n'", sentence=3383, chars=[0,204], words=[0,34])`

**Its parent Sentence's text:**	Previously, some basic machine learning classification algorithms have been used to evaluate the diagnostic power of noted antigens, but the issue of which algorithms perform best on this type of data is 


**60/104 Candidate/Span:**	`Span("b'This paper is motivated by a simple question regarding the diagnosis of such attacks-is it possible to establish attack-causality through network-level monitoring, without relying on signatures and attack-specific properties?'", sentence=2270, chars=[0,224], words=[0,38])`

**Its parent Sentence's text:**	This paper is motivated by a simple question regarding the diagnosis of such attacks-is it possible to establish attack-causality through network-level monitoring, without relying on signatures and attack-specific properties?

**61/104 Candidate/Span:**	`Span("b'For each program \xce\xb1, there is a box-modality [\xce\xb1] and a diamond modality\xe2\x8c\xa9 \xce\xb1\xe2\x8c\xaa. PDL was developed from first-order dynamic logic by Fischer-Ladner [FL79] and has become popular recently [GW09].'", sentence=1163, chars=[0,188], words=[0,43])`

**Its parent Sentence's text:**	For each program α, there is a box-modality [α] and a diamond modality〈 α〉. PDL was developed from first-order dynamic logic by Fischer-Ladner [FL79] and has become popular recently [GW09].

**62/104 Candidate/Span:**	`Span("b'Abstract: Sparse representation technique has been widely used in various areas of computer vision over the last decades.'", sentence=5179, chars=[0,120], words=[0,19])`

**Its parent Sentence's text:**	Abstract: Sparse representation technique has been widely used in various areas of computer vision over the last decades.

**63/104 Candidate/Span:**	`Span("b'We have extended previous work to consider robustness.'", sentence=4460, chars=[0,53], words=[0,8])`

**Its parent Sentence's text:**	We have extended previous work to consider robustness.

**64/104 Candidate/Span:**	`Span("b'In our previous work, we have automatically generated hints for logic tutoring by constructing a Markov Decision Process (MDP) that holds and rates historical student work for automatic selection of the best prior cases for hint generation.'", sentence=3197, chars=[0,239], words=[0,40])`

**Its parent Sentence's text:**	In our previous work, we have automatically generated hints for logic tutoring by constructing a Markov Decision Process (MDP) that holds and rates historical student work for automatic selection of the best prior cases for hint generation.

**65/104 Candidate/Span:**	`Span("b'Our system is faster than previous work by over an order of magnitude and it is capable of dealing with hundreds of millions of documents and thousands of topics.'", sentence=1728, chars=[0,161], words=[0,29])`

**Its parent Sentence's text:**	Our system is faster than previous work by over an order of magnitude and it is capable of dealing with hundreds of millions of documents and thousands of topics.

**66/104 Candidate/Span:**	`Span("b'Abstract Despite the recent trend of increasingly large datasets for object detection, there still exist many classes with few training examples.'", sentence=4830, chars=[0,144], words=[0,22])`

**Its parent Sentence's text:**	Abstract Despite the recent trend of increasingly large datasets for object detection, there still exist many classes with few training examples.

**67/104 Candidate/Span:**	`Span("b'Abstract: A technique for steering of flexible bevel-tipped needles through tissue, providing proportional control of trajectory curvature by means of duty-cycled rotation or spinning during insertion, has been presented previously, and tested in vitro in gelatin samples.'", sentence=4722, chars=[0,271], words=[0,45])`

**Its parent Sentence's text:**	Abstract: A technique for steering of flexible bevel-tipped needles through tissue, providing proportional control of trajectory curvature by means of duty-cycled rotation or spinning during insertion, has been presented previously, and tested in vitro in gelatin samples.

**68/104 Candidate/Span:**	`Span("b'The goal is object recognition, and three approaches are presented for distinguishing amongst objects from a previously-encountered set: The first approach is entirely geometric in nature and borrows ideas from "\n'", sentence=4197, chars=[0,212], words=[0,35])`

**Its parent Sentence's text:**	The goal is object recognition, and three approaches are presented for distinguishing amongst objects from a previously-encountered set: The first approach is entirely geometric in nature and borrows ideas from "


**69/104 Candidate/Span:**	`Span("b"In recent years, research attention in the software engineering community has shifted from process management and workflow tools that aim to plan for all coordination activity and eventualities before development begins to a new generation of more flexible tools that saturate the developer's workspace with information at varying degrees of granularity and in different visual, and often interactive, representations."", sentence=4992, chars=[0,417], words=[0,63])`

**Its parent Sentence's text:**	In recent years, research attention in the software engineering community has shifted from process management and workflow tools that aim to plan for all coordination activity and eventualities before development begins to a new generation of more flexible tools that saturate the developer's workspace with information at varying degrees of granularity and in different visual, and often interactive, representations.

**70/104 Candidate/Span:**	`Span("b'BackgrOuNd Considerable attention has been given recently to the online offerings such as edX and Coursera by prestigious universities such as MIT, Stanford, and Harvard.'", sentence=4347, chars=[0,169], words=[0,27])`

**Its parent Sentence's text:**	BackgrOuNd Considerable attention has been given recently to the online offerings such as edX and Coursera by prestigious universities such as MIT, Stanford, and Harvard.

**71/104 Candidate/Span:**	`Span("b'We show that none of the previous coalition structure generation algorithms can establish any \n'", sentence=1806, chars=[0,94], words=[0,14])`

**Its parent Sentence's text:**	We show that none of the previous coalition structure generation algorithms can establish any 


**72/104 Candidate/Span:**	`Span("b'We show that none of the previous coalition structure generation algorithms can establish any bound because \n'", sentence=2621, chars=[0,108], words=[0,16])`

**Its parent Sentence's text:**	We show that none of the previous coalition structure generation algorithms can establish any bound because 


**73/104 Candidate/Span:**	`Span("b'In previous work, we developed a reliable navigation technique that uses partially observable Markov models to represent metric, actuator and sensor uncertainties.'", sentence=5091, chars=[0,162], words=[0,24])`

**Its parent Sentence's text:**	In previous work, we developed a reliable navigation technique that uses partially observable Markov models to represent metric, actuator and sensor uncertainties.

**74/104 Candidate/Span:**	`Span("b'The variable circular plot (VCP) method for censusing animals is used widely to estimate the abundance of forest birds, despite a lack of effective methods for analyzing such data.'", sentence=5124, chars=[0,179], words=[0,32])`

**Its parent Sentence's text:**	The variable circular plot (VCP) method for censusing animals is used widely to estimate the abundance of forest birds, despite a lack of effective methods for analyzing such data.

**75/104 Candidate/Span:**	`Span("b'We develop a unified, discriminative model for such context-based action recognition building on recent techniques for learning large-scale discriminative models.'", sentence=4127, chars=[0,161], words=[0,25])`

**Its parent Sentence's text:**	We develop a unified, discriminative model for such context-based action recognition building on recent techniques for learning large-scale discriminative models.

**76/104 Candidate/Span:**	`Span("b'Abstract\xe2\x80\x94Session types are widely accepted as a useful expressive discipline for structuring communications in concurrent and distributed systems.'", sentence=3217, chars=[0,145], words=[0,20])`

**Its parent Sentence's text:**	Abstract—Session types are widely accepted as a useful expressive discipline for structuring communications in concurrent and distributed systems.

**77/104 Candidate/Span:**	`Span("b'Abstract In this work, we present a new method for doing object recognition using tactile force sensors that makes use of recent work on \xe2\x80\x9ctactile appearance\xe2\x80\x9d to describe objects by the spatially-varying appearance characteristics of their surface texture.'", sentence=4192, chars=[0,254], words=[0,43])`

**Its parent Sentence's text:**	Abstract In this work, we present a new method for doing object recognition using tactile force sensors that makes use of recent work on “tactile appearance” to describe objects by the spatially-varying appearance characteristics of their surface texture.

**78/104 Candidate/Span:**	`Span("b'As described in the previous chapters, some of the most successful work has included road following using CCD cameras [8.10] and cross country navigation using a scanning laser range finder [8.7].'", sentence=4705, chars=[0,195], words=[0,36])`

**Its parent Sentence's text:**	As described in the previous chapters, some of the most successful work has included road following using CCD cameras [8.10] and cross country navigation using a scanning laser range finder [8.7].

**79/104 Candidate/Span:**	`Span("b'Our recently proposed framework of deploying packet level redundancy elimination (RE) on network elements (eg, routers)([2],[3]) can serve as an effective way to improve network efficiency without expensive upgrades.'", sentence=1694, chars=[0,215], words=[0,35])`

**Its parent Sentence's text:**	Our recently proposed framework of deploying packet level redundancy elimination (RE) on network elements (eg, routers)([2],[3]) can serve as an effective way to improve network efficiency without expensive upgrades.

**80/104 Candidate/Span:**	`Span("b"Abstract The key role that video quality plays in impacting user engagement, and consequently providers' revenues, has motivated recent efforts in improving the quality of Internet video."", sentence=5341, chars=[0,186], words=[0,30])`

**Its parent Sentence's text:**	Abstract The key role that video quality plays in impacting user engagement, and consequently providers' revenues, has motivated recent efforts in improving the quality of Internet video.

**81/104 Candidate/Span:**	`Span("b'Abstract Recent biological advances with protein microarrays allow high-throughput antibody response screens against the entire proteome of a disease pathogen.'", sentence=3381, chars=[0,158], words=[0,22])`

**Its parent Sentence's text:**	Abstract Recent biological advances with protein microarrays allow high-throughput antibody response screens against the entire proteome of a disease pathogen.

**82/104 Candidate/Span:**	`Span("b'Our approach relies on recent learning theory results for distribution regression, using kernel \n'", sentence=2307, chars=[0,96], words=[0,14])`

**Its parent Sentence's text:**	Our approach relies on recent learning theory results for distribution regression, using kernel 


**83/104 Candidate/Span:**	`Span("b'Abstract Recent studies have shown a range of co-residency side channels that can be used to extract private information from cloud clients.'", sentence=2272, chars=[0,139], words=[0,24])`

**Its parent Sentence's text:**	Abstract Recent studies have shown a range of co-residency side channels that can be used to extract private information from cloud clients.

**84/104 Candidate/Span:**	`Span("b'Previous work with the programming language Lollimon demonstrates the expressive power of logic programming with linear logic in describing algorithms that have imperative elements or that must repeatedly make mutually exclusive \n'", sentence=5441, chars=[0,229], words=[0,31])`

**Its parent Sentence's text:**	Previous work with the programming language Lollimon demonstrates the expressive power of logic programming with linear logic in describing algorithms that have imperative elements or that must repeatedly make mutually exclusive 


**85/104 Candidate/Span:**	`Span("b'Despite the recent advances in full-body pose estimation using Kinect-like sensors, reliable monocular hand pose estimation in RGB-D images is still an unsolved problem.'", sentence=1617, chars=[0,168], words=[0,31])`

**Its parent Sentence's text:**	Despite the recent advances in full-body pose estimation using Kinect-like sensors, reliable monocular hand pose estimation in RGB-D images is still an unsolved problem.

**86/104 Candidate/Span:**	`Span("b'As described in our previous work, we developed Ariadne, a dependency plug-in infrastructure for the Eclipse IDE.'", sentence=3885, chars=[0,112], words=[0,21])`

**Its parent Sentence's text:**	As described in our previous work, we developed Ariadne, a dependency plug-in infrastructure for the Eclipse IDE.

**87/104 Candidate/Span:**	`Span("b'Abstract We review progress in a recent line of research that provides a concurrent computational interpretation of (intuitionistic) linear logic.'", sentence=3732, chars=[0,145], words=[0,22])`

**Its parent Sentence's text:**	Abstract We review progress in a recent line of research that provides a concurrent computational interpretation of (intuitionistic) linear logic.

**88/104 Candidate/Span:**	`Span("b'<\\nPrevious; Next >; Home > Conferences_Events > SCFORUM > 6.'", sentence=2902, chars=[0,60], words=[0,13])`

**Its parent Sentence's text:**	<\nPrevious; Next >; Home > Conferences_Events > SCFORUM > 6.

**89/104 Candidate/Span:**	`Span("b'In addition, previous negotiations are a source of heuristic \n'", sentence=4770, chars=[0,61], words=[0,10])`

**Its parent Sentence's text:**	In addition, previous negotiations are a source of heuristic 


**90/104 Candidate/Span:**	`Span("b'Abstract Recent studies have investigated how a team of mobile sensors can cope with real world constraints, such as uncertainty in the reward functions, dynamically appearing and disappearing targets, technology failures end changes in the environment conditions.'", sentence=4064, chars=[0,263], words=[0,40])`

**Its parent Sentence's text:**	Abstract Recent studies have investigated how a team of mobile sensors can cope with real world constraints, such as uncertainty in the reward functions, dynamically appearing and disappearing targets, technology failures end changes in the environment conditions.

**91/104 Candidate/Span:**	`Span("b'Recent results make a promising case for augmenting an oversubscribed network with reconfigurable inter-rack wireless or optical links.'", sentence=2288, chars=[0,134], words=[0,20])`

**Its parent Sentence's text:**	Recent results make a promising case for augmenting an oversubscribed network with reconfigurable inter-rack wireless or optical links.

**92/104 Candidate/Span:**	`Span("b'In this setting, a key challenge to high availability is the presence of correlated failures, not addressed by previous monitoring services.'", sentence=2008, chars=[0,139], words=[0,23])`

**Its parent Sentence's text:**	In this setting, a key challenge to high availability is the presence of correlated failures, not addressed by previous monitoring services.

**93/104 Candidate/Span:**	`Span("b'In this paper, we investigate how to use machine learning to add reactivity to a previously \n'", sentence=2476, chars=[0,92], words=[0,17])`

**Its parent Sentence's text:**	In this paper, we investigate how to use machine learning to add reactivity to a previously 


**94/104 Candidate/Span:**	`Span("b'Network-awareness is a recent attempt to bridge the gap between the realities of networks and the demands of applications.'", sentence=4814, chars=[0,121], words=[0,21])`

**Its parent Sentence's text:**	Network-awareness is a recent attempt to bridge the gap between the realities of networks and the demands of applications.

**95/104 Candidate/Span:**	`Span("b'As one of the most active research methods in the field of data mining, the data mining method of association rule has been widely used in various aspects in the field of traditional Chinese medicine, which makes the large database of traditional Chinese medicine information can be utilized effectively, and promotes the development of traditional Chinese medicine modernization.'", sentence=4936, chars=[0,379], words=[0,61])`

**Its parent Sentence's text:**	As one of the most active research methods in the field of data mining, the data mining method of association rule has been widely used in various aspects in the field of traditional Chinese medicine, which makes the large database of traditional Chinese medicine information can be utilized effectively, and promotes the development of traditional Chinese medicine modernization.

**96/104 Candidate/Span:**	`Span("b'Our results are for a class of games that generalizes the only previously known class of imperfect-recall abstractions where any results had been obtained.'", sentence=1585, chars=[0,154], words=[0,26])`

**Its parent Sentence's text:**	Our results are for a class of games that generalizes the only previously known class of imperfect-recall abstractions where any results had been obtained.

**97/104 Candidate/Span:**	`Span("b'Abstract Recently, there has been increasing deployment of distributed infrastructure and applications such as Akamai, PlanetLab and DHTs.'", sentence=2006, chars=[0,137], words=[0,20])`

**Its parent Sentence's text:**	Abstract Recently, there has been increasing deployment of distributed infrastructure and applications such as Akamai, PlanetLab and DHTs.

**98/104 Candidate/Span:**	`Span("b'The robot design extends upon previous prototypes of HeartLander, a miniature mobile robot that moves in an inchworm-like fashion.'", sentence=4718, chars=[0,129], words=[0,22])`

**Its parent Sentence's text:**	The robot design extends upon previous prototypes of HeartLander, a miniature mobile robot that moves in an inchworm-like fashion.

**99/104 Candidate/Span:**	`Span("b'Continuous lattices, as the authors have learned in recent years, exhibit a variety of different aspects, some are lattice theoretical, some are topological, some belong to topological algebra and some to category theory'", sentence=2126, chars=[0,219], words=[0,37])`

**Its parent Sentence's text:**	Continuous lattices, as the authors have learned in recent years, exhibit a variety of different aspects, some are lattice theoretical, some are topological, some belong to topological algebra and some to category theory

**100/104 Candidate/Span:**	`Span("b'Abstract: Switched LANs are become more widely used because they can provide a higher bandwidth than LANs based on shared media.'", sentence=1211, chars=[0,127], words=[0,22])`

**Its parent Sentence's text:**	Abstract: Switched LANs are become more widely used because they can provide a higher bandwidth than LANs based on shared media.

**101/104 Candidate/Span:**	`Span("b'As sketched in the previous lecture, this an important piece in the general technique of verifying properties of logic programs by reasoning about proof terms.\n'", sentence=1077, chars=[0,159], words=[0,27])`

**Its parent Sentence's text:**	As sketched in the previous lecture, this an important piece in the general technique of verifying properties of logic programs by reasoning about proof terms.


**102/104 Candidate/Span:**	`Span("b'In recent years, industry has developed a variety of standards, eg SOAP, WSDL, UDDI, BPEL for web services discovery, description, and distributed execution over the Web.'", sentence=4972, chars=[0,169], words=[0,33])`

**Its parent Sentence's text:**	In recent years, industry has developed a variety of standards, eg SOAP, WSDL, UDDI, BPEL for web services discovery, description, and distributed execution over the Web.

**103/104 Candidate/Span:**	`Span("b'ACT-R is one of the most widely used cognitive architectures, and it has been used to model hundreds of phenomena described in the cognitive psychology literature.'", sentence=4315, chars=[0,162], words=[0,29])`

**Its parent Sentence's text:**	ACT-R is one of the most widely used cognitive architectures, and it has been used to model hundreds of phenomena described in the cognitive psychology literature.


Check if any of the extracted candidates come from the same sentence? False


**(Solved) Question 1:** the `CandidateExtractor` extracts only spans with length 1. But Each sentence is only getting matched once with one span (try search "sentence=3993", for example). The sentence is good, but we still would want longer span, e.g. half part or the whole of a sentence. 

`Span("b'previous'", sentence=4705, chars=[20,27], words=[4,4])`

**Answer 1:** overwrite `DictionaryMatch._f()`.

=====================================================================================

**(Solved) Question 2:** Several `Span` corresponds to the same sentence, e,g, three `Span` corresponds to `sentence=490`. They wee all included even after we have set `longest_match_only=True`, as these spans all have the longest length.  

**Answer 2:** Perhaps try either (1) filtering by first appearance of document ID; (2) set max_length as the length of the longest sentence length.


### Stopped here as of Wed 06/19 9:39 pm. More to come ... ###

## TODO List

1. Re-format 2K gold label papers. Load them as gold standard. Question: are there any recommended approaches to evaluate segmentation quality, e.g. word overlap? 
2. Our current segmentation strictly segments by sentence boundary, is that errorenous? 



Thanks for reading! 

Some debugging note to memorize (could ignore). 

```
python -m spacy download en
```

Current issue: parser does not parse *by periods*. Sentence count is significantly fewer than expected! 
Potential fix: https://github.com/explosion/spaCy/issues/93

======= Some more debugging log here (not necessary, could skip reading) ======
~~~~
Xins-MacBook-Pro:~ xin$ source activate snorkel
(snorkel) Xins-MacBook-Pro:~ xin$ python
Python 3.6.4 |Anaconda custom (64-bit)| (default, Jan 16 2018, 12:04:33) 
[GCC 4.2.1 Compatible Clang 4.0.1 (tags/RELEASE_401/final)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import spacey
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'spacey'
>>> import spacy
>>> spacy.load('en')
<spacy.en.English object at 0x1080e1da0>
>>> model=spacy.load('en')
>>> docs=model.tokenizer('Hello, world. Here are two sentences.')
>>> for sent in docs.sents:
...     pritn(sent.text)
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "spacy/tokens/doc.pyx", line 439, in __get__ (spacy/tokens/doc.cpp:9808)
ValueError: Sentence boundary detection requires the dependency parse, which requires data to be installed. For more info, see the documentation: 
https://spacy.io/docs/usage

>>> for sent in docs.sents:
...     print(sent.text)
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "spacy/tokens/doc.pyx", line 439, in __get__ (spacy/tokens/doc.cpp:9808)
ValueError: Sentence boundary detection requires the dependency parse, which requires data to be installed. For more info, see the documentation: 
https://spacy.io/docs/usage

>>> from spacy.en import English
>>> nlp = English()
>>> doc = nlp(raw_text)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'raw_text' is not defined
>>> raw_text='Hello, world. Here are two sentences.'
>>> doc = nlp(raw_text)
>>> sentences = [sent.string.strip() for sent in doc.sents]
>>> sentences
['Hello, world.', 'Here are two sentences.']
>>> model(raw_text)
Hello, world. Here are two sentences.
>>> docs=model(raw_text)
>>> docs.sents
<generator object at 0x14ad31948>
>>> docs=model(raw_text)
>>> sentences = [sent.string.strip() for sent in docs.sents]
>>> sentences
['Hello, world.', 'Here are two sentences.']
>>> 
~~~~
